lib/util/cache.js

```diff
@@ -0,0 +1,63 @@
+/**
+ * @fileoverview Cache object that manages multiple namespaced collections of data.
+ * @author Kai Cataldo
+ */
+
+"use strict";
+
+//------------------------------------------------------------------------------
+// Public Interface
+//------------------------------------------------------------------------------
+
+/**
+ * Cache
+ *
+ * A cache object that manages multiple namespaced collections of data.
+ * Each namespaced collection is a WeakMap instance.
+ * The cache will lazily create a new cache for the specified collection namespace the first
+ * time set() is called with a new collection type.
+ */
+module.exports = class Cache {
```

Comment:
Good call - I'm honestly not sold on implementing caching right now (especially if we move these methods into the TokenStore and it's more performant), since running `npm run perf` didn't show consistently faster results with caching vs non-caching. 

I also think that caching is now less impactful because we're not calculating both leading and trailing comments (unnecessarily, most of the time) for a node any more. With `sourceCode.getComments()` there was a higher chance the same cached result would be reused.

Seems to me like this is something we could hold off on for now and implement in the future if we see the need/feel like there are big perf gains to be had as we optimize 4.0.